### ISSUE TITLE
cloud_firestore difficulty with pagination

### ISSUE THREAD

**Issue by jinfeei** [flutter/flutter#21017]
startAfter (for dart/flutter) currently does not support using "DocumentSnapshot" as its parameter. Without this, I find it difficult to implement pagination. Refer the following snippet from firestore documentation (for Web). How can I achieve the same using Dart/Flutter?
```dart
var first = db.collection("cities")
        .orderBy("population")
        .limit(25);

return first.get().then(function (documentSnapshots) {
  // Get the last visible document
  var lastVisible = documentSnapshots.docs[documentSnapshots.docs.length-1];
  console.log("last", lastVisible);

  // Construct a new query starting at this document,
  // get the next 25 cities.
  var next = db.collection("cities")
          .orderBy("population")
          .startAfter(lastVisible)
          .limit(25);
});
```

**Comment by naumanahmed19** [comment]
@jinfeei  have you found any solution?

**Comment by zoechi** [comment]
Please consider asking support questions in one of the other channels listed at http://flutter.io/support .

**Comment by creativecreatorormaybenot** [comment]
It should work fine just using the values from the document. If you have the `DocumentSnapshot`, you will also have access to the values: https://stackoverflow.com/a/54660349/6509751

**Comment by collinjackson** [comment]
I think it would be challenging to implement passing of `DocumentSnapshot` to the native side of the plugin. Because there's no public `DocumentSnapshot` constructor API, we'd have to hang on to a copy of every `DocumentSnapshot` we ever got from the native SDK just in case we need it later, and there's no notification that would let us know that the `DocumentSnapshot` reference could be released, so we'd leak memory.

It seems to me that the [approach](https://stackoverflow.com/questions/54660348/pagination-in-cloud-firestore/54660349#54660349) @creativecreatorormaybenot is suggesting can be used implement pagination, and this is the best we can hope for in the short term until cloud_firestore is rewritten as a pure Dart package. @jinfeei can you let em know if this addresses your issue?

**Comment by cokn-s** [comment]
I am having the same difficulty as @jinfeei and while the solution @creativecreatorormaybenot proposed of just using values in the startAfter field does work if the field you are looking at is completely unique it does not work for the case where some documents have the same value in that field.
see video if this does not make sense: 
[https://www.youtube.com/watch?v=poqTHxtDXwU&feature=youtu.be](url)

So basically what is happening for me is that I am skipping a bunch of documents that have the same field value using pagination.  Firestore's solution for this is to use the document snapshot, do we have a timeline on when this might be accomplished?

**Comment by creativecreatorormaybenot** [comment]
@cokn-s You can use `.orderBy(...).orderBy('__name__').startAfter([..., lastVisible.documentId])`. Where is my award?

**Comment by cokn-s** [comment]
@creativecreatorormaybenot Thanks a bunch!  I had to add an index but other than that this worked perfectly!

**Comment by kroikie** [comment]
@jinfeei

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/690. Any further collaboration will be done there.

**Comment by excogitatr** [comment]
Try this package [paginate_firestore](https://pub.dev/packages/paginate_firestore). You can simply just pass the firestore query and it will handle all the `startAfterDocument` part.
